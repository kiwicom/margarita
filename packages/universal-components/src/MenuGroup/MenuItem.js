// @flow

import * as React from 'react';
import { View, Platform } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import Separator from './Separator';
import { Icon } from '../Icon';
import { StyleSheet, type StylePropType } from '../PlatformStyleSheet';
import { Text } from '../Text';
import { Touchable } from '../Touchable';

type Props = {|
  +title: React.Node,
  +onPress: () => void,

  +titleStyle?: StylePropType,
  +icon?: React.Element<typeof Icon>,
  +subTitle?: React.Node,
  +subTitleStyle?: StylePropType,
  +label?: React.Node,
  +labelStyle?: StylePropType,
  +hideActionIcon?: boolean,
|};

export default function MenuItem({
  title,
  onPress,
  icon,
  titleStyle,
  subTitle,
  subTitleStyle,
  label,
  labelStyle,
  hideActionIcon = false,
}: Props) {
  const menuIcon =
    icon != null
      ? React.cloneElement(icon, {
          color: icon.props.color ?? defaultTokens.paletteInkLight,
          size: icon.props.size ?? 'medium',
          style: [styles.icon, icon.props.style],
        })
      : null;
  return (
    <View style={styles.container}>
      <Touchable onPress={onPress} style={styles.wrapper}>
        <View style={styles.row}>
          {menuIcon}
          <View>
            {label != null && (
              <Text type="secondary" size="small" style={labelStyle}>
                {label}
              </Text>
            )}
            <Text type="attention" size="large" style={titleStyle}>
              {title}
            </Text>
            {subTitle != null && (
              <Text type="secondary" size="small" style={subTitleStyle}>
                {subTitle}
              </Text>
            )}
          </View>
          {Platform.OS === 'ios' && hideActionIcon === false && (
            <View style={styles.actionIcon}>
              <Icon
                name="chevron-right"
                color={defaultTokens.paletteProductNormal}
              />
            </View>
          )}
        </View>
      </Touchable>
      <Separator style={icon == null ? styles.noIcon : styles.iconSeparator} />
    </View>
  );
}

const wrapperPaddingStart = 12;
const iconWidth = 24;
const iconMargin = 12;

const styles = StyleSheet.create({
  container: {
    marginTop: StyleSheet.hairlineWidth,
    ios: {
      marginBottom: -StyleSheet.hairlineWidth,
    },
  },
  wrapper: {
    backgroundColor: defaultTokens.paletteWhite,
    ios: {
      paddingTop: 8.5,
      paddingBottom: 11.5,
    },
    paddingTop: 6,
    paddingBottom: 8.5,

    paddingStart: wrapperPaddingStart,
    paddingEnd: 16,
  },
  noIcon: {
    ios: {
      marginStart: 12,
    },
  },
  iconSeparator: {
    ios: {
      marginStart: wrapperPaddingStart + iconWidth + iconMargin,
    },
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  icon: {
    marginEnd: iconMargin,
  },
  actionIcon: {
    flex: 1,
    alignItems: 'flex-end',
  },
});
