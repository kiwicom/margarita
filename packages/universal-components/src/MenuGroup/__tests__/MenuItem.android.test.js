// @flow

import * as React from 'react';
import { shallow } from 'react-native-testing-library';

import MenuItem from '../MenuItem';

jest.mock('Platform', () => ({
  OS: 'android',
}));

describe('MenuGroup', () => {
  it('renders correctly', () => {
    const { output } = shallow(<MenuItem onPress={jest.fn()} title="test" />);
    expect(output).toMatchSnapshot();
  });
});
