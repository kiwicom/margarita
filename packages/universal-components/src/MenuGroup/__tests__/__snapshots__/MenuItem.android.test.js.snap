// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuGroup renders correctly 1`] = `
<Component
  style={
    Object {
      "marginTop": 0.5,
    }
  }
>
  <Touchable
    accessibilityComponentType="button"
    accessibilityTraits="button"
    borderlessRipple={false}
    disabled={false}
    onPress={[MockFunction]}
    rippleColor="rgba(0, 0, 0, 0.32)"
  >
    <MenuItemWrapper
      iconRendered={false}
    >
      <Component
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Component>
          <WithTheme
            size="large"
            type="attention"
          >
            test
          </WithTheme>
        </Component>
      </Component>
    </MenuItemWrapper>
  </Touchable>
</Component>
`;
